lib/bookingsync/engine/account_model.rb

```diff
@@ -0,0 +1,114 @@
+class BookingsyncEngineUniquenessValidator < ActiveRecord::Validations::UniquenessValidator
+  def initialize(options)
+    if options[:scope].present? && options[:scope].respond_to?(:call)
+      options[:scope] = options[:scope].call
+    end
+    super
+  end
+end
+
+module BookingSync::Engine::AccountModel
+  extend ActiveSupport::Concern
+
+  included do
+    validates :synced_id, bookingsync_engine_uniqueness: {
+      scope: Proc.new { BookingSync::Engine.support_multi_applications? ? :host : nil }
+    }
+    scope :authorized, -> { where.not(oauth_access_token: nil) }
+  end
+
+  module ClassMethods
+    def from_omniauth(auth, host: nil)
+      if BookingSync::Engine.support_multi_applications?
+        if host.present?
+          account = find_or_initialize_by(host: host, synced_id: auth.uid, provider: auth.provider)
+        else
+          raise ArgumentError, "The `host` variable must be passed when using BookingSync Engine with " \
+                               "multi application support"
+        end
+      else
+        account = find_or_initialize_by(synced_id: auth.uid, provider: auth.provider)
+      end
+
+      account.tap do |account|
+        account.name = auth.info.business_name
+        account.update_token(auth.credentials)
+        account.save!
+      end
+    end
+  end
+
+  def token
+    @token ||= begin
+      token_options = {}
+      if oauth_refresh_token
+        token_options[:refresh_token] = oauth_refresh_token
+        token_options[:expires_at]    = oauth_expires_at
+      end
+
+      token = OAuth2::AccessToken.new(oauth_client, oauth_access_token, token_options)
+
+      if token.expired?
+        refresh_token!(token)
+      else
+        token
+      end
+    end
+  end
+
+  def application_token
+    if BookingSync::Engine.support_multi_applications?
+      BookingSync::Engine.application_token(
+        client_id: application.client_id,
+        client_secret: application.client_secret
+      )
+    else
+      BookingSync::Engine.application_token
+    end
+  end
+
+  def oauth_client
+    if BookingSync::Engine.support_multi_applications?
+      BookingSync::Engine.oauth_client(
+        client_id: application.client_id,
+        client_secret: application.client_secret
+      )
+    else
+      BookingSync::Engine.oauth_client
+    end
+  end
+
+  def application
+    if BookingSync::Engine.support_multi_applications?
```

Comment:
All these `if BookingSync::Engine.support_multi_applications?` look like something is off here.

How about having two separate modules: AccountModel and AccountModelForMultiApplicationSupport? That way we would actually solve the issue I mentioned about `from_omniauth`. Polymorphism FTW.

Reply:
This goes way further than just the model. This changes are in the Engine, Helpers, ... as well.

An option could be to have 2 engines actually, yet feels like a lot of duplication and not sure it's the best choice either.

Reply:
This change would be super useful. Indeed, there are similar things in Engine, Helpers, but they are less drastic than here, one conditional is not bad and maybe not worth changing there, but given that there are multiple conditionals for this module, this would be a definitely change in the right direction